### PR TITLE
Initial SearchUserModal implementation

### LINKED ISSUE #59: Add Collaborators to BookLists Implementation
- Add Frontend Design so Collaborators can access and view other users BookLists.
- Show a list of Collaborators for a selected BookList

### LINKED ISSUE #86: Friend Functionality
Allow users to add friends via UserCards, see what their friends are reading in the sidebar feed, share lists with friends, invite friends to clubs, etc.

### PR DESCRIPTION
**Pull request by MauroGuerrero20**
# Features
- Implements the ```UserSearchServlet``` 
   - Utilizes prefix matching to search for Users on Firestore based on a search term send from the Frontend. After finishing its search, it then sends back the search results to the Frontend.
- Adds a user search modal (```SearchUserModal.jsx```)
   - Allows users to search for other users. 
  - Utilizes the ```UserSearchServlet``` to send a search request (GET). 
  - Displays search results if any.
- No Firestore Integration with BookLists, Clubs, or Friends yet (For future PRs).
  - BookList Collaborators
  - Clubs Members
  - Friends

# Demo
User Modal Frontend Functionality Demo

### CODE REVIEW

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (antoniolinhart) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: nit: This comment looks like it's from your other class

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (antoniolinhart) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java:60]: This is a good way to ensure the sensitive token doesn't get passed to the frontend where the end user can potentially see it

File: capstone-backend/src/main/java/com/google/sps/servlets/UserServlet.java
Comment (antoniolinhart) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserServlet.java:112]: If the userid the user is asking for doesn't exist, it might be helpful to return a bad request so they know what happened

File: capstone-frontend/src/components/Home.js
Comment (antoniolinhart) [inline capstone-frontend/src/components/Home.js]: Why is this on the home page?

File: capstone-frontend/src/components/SearchBookModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchBookModal.jsx:221]: These button tags don't line up

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchUserModal.jsx]: nit: Stay consistent with camel case naming convention `indexAddedUsersTracker`

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchUserModal.jsx:94]: It might be better to make a deep copy of this array to avoid changing the state without set state at the end? I know that you have a setState at the end of the function but it's generally bad practice to change a state variable like this

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchUserModal.jsx]: Same comments as above

Comment (antoniolinhart) [reply capstone-frontend/src/components/SearchUserModal.jsx]: Also you never properly update the addedUsersTracker in this function

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchUserModal.jsx:205]: You should make the buttons have the same width by using something like w-50

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchUserModal.jsx:240]: nit: same as above

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchUserModal.jsx]: nit: empty space at the end of the classname string literal

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (antoniolinhart) [inline capstone-frontend/src/components/SearchUserModal.jsx:174]: Could these user cards be abstracted and used for my User Search page too?

**Review by antoniolinhart — APPROVED** [review]
Looks good to me!

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (MauroGuerrero20) [reply capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: Changed.

File: capstone-frontend/src/components/Home.js
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/Home.js]: To test the UserSearchModal for now. Will remove it in a future PR.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx]: Done.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx:174]: I am not sure. Probably they are not that proprietary to this modal.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx]: Ok

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx]: Nice Catch. Changed

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx:205]: I will change the width in the next PR.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserServlet.java
Comment (MauroGuerrero20) [reply capstone-backend/src/main/java/com/google/sps/servlets/UserServlet.java:112]: If an error is throw here, then it because of a Firebase issue. I added a ```response.sendError``` to the catch.

File: capstone-frontend/src/components/SearchBookModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchBookModal.jsx:221]: Changed.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx:94]: Will try to change in the next PR.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (steven-solar) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java:50]: What is the use of the escape character here?

**Review by steven-solar — APPROVED** [review]
This will be great for adding users to lists and clubs! Seems like one of the last missing pieces for our demo.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (MauroGuerrero20) [reply capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java:50]: Used for prefix matching.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (VivekShankar) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: nit: "No Parameters Specified"

File: capstone-backend/src/main/java/com/google/sps/servlets/UserServlet.java
Comment (VivekShankar) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserServlet.java:108]: Good job restricting the `try` `catch` to the particular section of the code that could throw exceptions, and also catching the specific exceptions that could be thrown!

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (VivekShankar) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: This `if` condition here doesn't seem to match up with the "No parameters specified" error message. Based on that message, I think I'd expect the condition to be: `if (searchTerm.isEmpty() && request.getParameterMap().size() == 0);`?

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (VivekShankar) [inline capstone-frontend/src/components/SearchUserModal.jsx:19]: Just a couple of clarification questions here...
1. Is `addedUsersTracker` just a combination of `addedUsers` and `addedFriends` below? 
2. What is the distinction between an "addedUser" and an "addedFriend" ? When would a user choose to add a user vs. add a friend?

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (geoffbass) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java:21]: nit: /api/users might be a bit cleaner here

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (geoffbass) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: The error message here doesn't seem to apply to the `size() > 1` case. Can you split that out with its own debug output?

**Review by VivekShankar — APPROVED** [review]
LGTM

File: capstone-frontend/src/components/Home.js
Comment (geoffbass) [reply capstone-frontend/src/components/Home.js]: Is there a way to test it without integrating this change into the master branch? Any automated testing that can be done?

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [inline capstone-frontend/src/components/SearchUserModal.jsx:14]: Good descriptive comment

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [inline capstone-frontend/src/components/SearchUserModal.jsx]: nit: missing ;

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [inline capstone-frontend/src/components/SearchUserModal.jsx:42]: Does this mean the API response won't always have a payload? Could that be changed to send an empty array instead? Or is this just the case when there's an error?

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [inline capstone-frontend/src/components/SearchUserModal.jsx:50]: What do "Guilty" and "Not Guilty" mean in this context?

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [inline capstone-frontend/src/components/SearchUserModal.jsx:68]: nice debounce implementation

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [inline capstone-frontend/src/components/SearchUserModal.jsx:89]: What's the purpose of the `addedUsersTracker` in these methods? It looks like you could omit it and set `addedUsers` and `addedFriends` directly in the `if` cases.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [reply capstone-frontend/src/components/SearchUserModal.jsx:94]: You should be able to use a `filter` here to create a new array and then set the state to it

File: capstone-frontend/src/styles/SearchUserModal.css
Comment (geoffbass) [inline capstone-frontend/src/styles/SearchUserModal.css]: Try using an id selector here so you can get rid of `!important`

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (geoffbass) [inline capstone-frontend/src/components/SearchUserModal.jsx:122]: Generally with this much JSX you can split this up into a few sub-components, though I'll leave that to your judgment here

**Review by geoffbass — CHANGES_REQUESTED** [review]
Just a few things to change before merging this, ping me once you've made the updates so I'm not blocking you

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (MauroGuerrero20) [reply capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: Correct, I will change the implementation.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (MauroGuerrero20) [reply capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: Will do.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx:19]: The ```addedUsersTracker``` keeps track of Users added to ```addedUsers``` and addedFriends``` and displays them in the modal. ```addedUsersTracker``` basically combines both of these array into a set to avoids displaying duplicate data. 

```addedUsers``` is meant to be used as a helper array. For example, I will use it to keep track of Collaborators. I think @steven-solar will use it to keep track of the members of the club.

```addedFriends``` is meant to keep track of the User's friends. This allows an owner/admin to not only add a User to Club but also add them as a Friend. 

In the future, we might change the Friend functionality of the modal.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (MauroGuerrero20) [reply capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java:21]: There is already a ```"/api/user"``` WebServlet so I want to differentiate this servlet to prevent errors.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx]: Done.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx:42]: The API might not find any users, I think at the moment it sends an empty array if that is the case. This check should only be true if an error occurs.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (MauroGuerrero20) [reply capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java]: Done.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx:50]: We assume that ```resultsFound``` is true. When we finally fetch the searchTerm, if it return an empty array then ```resultsFound``` is guilty and set to false. If ```resultsFound``` is false then we render ```No Results Found```.

File: capstone-frontend/src/components/SearchUserModal.jsx
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/SearchUserModal.jsx:89]: ```addedUsersTracker``` is a helper array I use to render the ```addedUsers``` or ```addedFriends``` arrays. Without it, I would have to render both arrays separately, which may render duplicate data. In these functions, ```addedUsersTracker``` is checking if the it contains the user being added. These check stop ```addedUsersTracker``` from adding the same user twice.

File: capstone-frontend/src/styles/SearchUserModal.css
Comment (MauroGuerrero20) [reply capstone-frontend/src/styles/SearchUserModal.css]: Done. I forgot that the id selector can override classes.

File: capstone-frontend/src/components/Home.js
Comment (MauroGuerrero20) [reply capstone-frontend/src/components/Home.js]: I don't really know how I would automate it. The modal require user input to test it. I removed it so I doesn't go into master.

File: capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java
Comment (geoffbass) [inline capstone-backend/src/main/java/com/google/sps/servlets/UserSearchServlet.java:40]: Yeah this looks better, thanks

**Review by geoffbass — APPROVED** [review]
LGTM

**Pull request merged**
